dmriprep/utils/vectors.py

```diff
@@ -1,482 +1,346 @@
+BVEC_NORM_EPSILON = 0.1
+
+
+class DiffusionGradientTable:
+    """Data structure for DWI gradients."""
+
+    __slots__ = ['_affine', '_gradients', '_b_scale', '_bvecs', '_bvals', '_normalized',
+                 '_b0_thres', '_bvec_norm_epsilon']
+
+    def __init__(self, dwi_file=None, bvecs=None, bvals=None, rasb_file=None,
+                 b_scale=True, b0_threshold=B0_THRESHOLD, bvec_norm_epsilon=BVEC_NORM_EPSILON):
+        """
+        Create a new table of diffusion gradients.
+
+        Parameters
+        ----------
+        dwi_file : str or os.pathlike or nibabel.spatialimage
+            File path to the diffusion-weighted image series to which the
+            bvecs/bvals correspond.
+        bvals : str or os.pathlike or numpy.ndarray
+            File path of the b-values.
+        bvecs : str or os.pathlike or numpy.ndarray
+            File path of the b-vectors.
+        rasb_file : str or os.pathlike
+            File path to a RAS-B gradient table. If rasb_file is provided,
+            then bvecs and bvals will be dismissed.
+        b_scale : bool
+            Whether b-values should be normalized.
+
+        """
+        self._b_scale = b_scale
+        self._b0_thres = b0_threshold
+        self._bvec_norm_epsilon = bvec_norm_epsilon
+        self._gradients = None
+        self._bvals = None
+        self._bvecs = None
+        self._affine = None
+        self._normalized = False
+
+        if dwi_file is not None:
+            self.affine = dwi_file
+        if rasb_file is not None:
+            self.gradients = rasb_file
+            self.generate_vecval()
+        elif dwi_file and bvecs and bvals:
+            self.bvecs = bvecs
+            self.bvals = bvals
+            self.generate_rasb()
+
+    @property
+    def affine(self):
+        return self._affine
+
+    @property
+    def gradients(self):
+        """Get gradient table (rasb)."""
+        return self._gradients
+
+    @property
+    def bvecs(self):
+        return self._bvecs
+
+    @property
+    def bvals(self):
+        return self._bvals
+
+    @property
+    def normalized(self):
+        return self._normalized
+
+    @affine.setter
+    def affine(self, value):
+        if isinstance(value, (str, Path)):
+            dwi_file = nb.load(str(value))
+            self._affine = dwi_file.affine.copy()
+            return
+        if hasattr(value, 'affine'):
+            self._affine = value.affine
+        self._affine = np.array(value)
+
+    @gradients.setter
+    def gradients(self, value):
+        if isinstance(value, (str, Path)):
+            value = np.loadtxt(value, skiprows=1)
+        self._gradients = value
+
+    @bvecs.setter
+    def bvecs(self, value):
+        if isinstance(value, (str, Path)):
+            value = np.loadtxt(str(value)).T
+        else:
+            value = np.array(value, dtype='float32')
+
+        # Correct any b0's in bvecs misstated as 10's.
+        value[np.any(abs(value) >= 10, axis=1)] = np.zeros(3)
```

Comment:
you had this setting [1, 0, 0] instead [0, 0, 0] - was that intended?

Reply:
So sort of. Basically the story behind this is that when I was working with NeuroData using the KKI dataset, we discovered that the B0 vectors were acquired orthogonally (and had some weird labeling such as [1, 0, 0], [0, 0, 1]). Often times we also encountered corrupted cases where there was a '10' in place of the 1, so we hard coded those cases to force to 1. But looking back, since they are actually B0's anyway, I think it's probably perfectly fine to force them to [0, 0, 0] no matter what given that they have no diffusion encoding that becomes relevant anyway. In other words, your correction is perfectly fine and probably more robust across datasets.